src/main/scala/cognite/spark/v1/EventsRelation.scala

```diff
@@ -76,8 +78,21 @@ class EventsRelation(config: RelationConfig)(val sqlContext: SQLContext)
   }
 
   override def update(rows: Seq[Row]): IO[Unit] = {
-    val events = rows.map(r => fromRow[EventUpdate](r))
-    client.events.update(events) *> IO.unit
+    val eventUpdates = rows.map(r => fromRow[EventsUpsertSchema](r))
+    val (updatesById, updatesByExternalId) = eventUpdates.partition(
+      p => p.id.isDefined
+    )
+    val id = if (updatesById.isEmpty) { IO.unit } else {
+      client.events.updateById(updatesById.map(ev => ev.id.get -> ev.transformInto[EventUpdate]).toMap)
+    }
+    val extId = if (updatesByExternalId.isEmpty) { IO.unit } else {
+      client.events.updateByExternalId(updatesByExternalId
+        .map(ev =>
+          ev.externalId
+            .getOrElse("") -> ev.into[EventUpdate].withFieldComputed(_.externalId, _ => None).transform)
```

Comment:
And again about `.getOrElse`. We seem to be doing this "partition, then map to `externalId -> ...into[EventUpdate].withFieldComputed(...)`" several times, so maybe we can move that into a utility function?